<a href="https://colab.research.google.com/github/IkutoYoshioka/AI-/blob/main/OLS%E6%8E%A8%E5%AE%9A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm

In [4]:
data = pd.read_csv('/content/drive/MyDrive/数理統計/データサイエンス課題/data_HW2 - data_HW2.csv')
data.head()

age  educ  birthwght  smoke  male
0   21    11       3200      0     0
1   31    14       3204      0     0
2   27    16       3175      0     1
3   34    17       4309      0     1
4   22    12       3529      0     1

## (a)

In [6]:
X = data["age"]
Y = data["birthwght"]

X_const = sm.add_constant(X)
model = sm.OLS(Y, X_const)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:              birthwght   R-squared:                       0.006
Model:                            OLS   Adj. R-squared:                  0.005
Method:                 Least Squares   F-statistic:                     5.672
Date:                Sat, 07 Dec 2024   Prob (F-statistic):             0.0174
Time:                        09:27:33   Log-Likelihood:                -7713.7
No. Observations:                1000   AIC:                         1.543e+04
Df Residuals:                     998   BIC:                         1.544e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3237.0604     91.603     35.338      0.0

## (b)

In [7]:
X = data[["age", "educ"]]
Y = data["birthwght"]

X_const = sm.add_constant(X)
model = sm.OLS(Y, X_const)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:              birthwght   R-squared:                       0.008
Model:                            OLS   Adj. R-squared:                  0.006
Method:                 Least Squares   F-statistic:                     3.928
Date:                Sat, 07 Dec 2024   Prob (F-statistic):             0.0200
Time:                        09:32:56   Log-Likelihood:                -7712.6
No. Observations:                1000   AIC:                         1.543e+04
Df Residuals:                     997   BIC:                         1.545e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3138.7617    113.226     27.721      0.0

In [8]:
# 不均一分散に頑強な標準誤差の算出

results2 = model.fit(cov_type='HC3')
print(results2.summary())

                            OLS Regression Results                            
Dep. Variable:              birthwght   R-squared:                       0.008
Model:                            OLS   Adj. R-squared:                  0.006
Method:                 Least Squares   F-statistic:                     3.523
Date:                Sat, 07 Dec 2024   Prob (F-statistic):             0.0299
Time:                        09:44:44   Log-Likelihood:                -7712.6
No. Observations:                1000   AIC:                         1.543e+04
Df Residuals:                     997   BIC:                         1.545e+04
Df Model:                           2                                         
Covariance Type:                  HC3                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const       3138.7617    119.799     26.200      0.0

## (c)

In [10]:
#1
X_full = sm.add_constant(data[["age", "educ"]])
y = data["birthwght"]
model_full = sm.OLS(y, X_full).fit()
beta1_full = model_full.params["age"]

#2
X_partial = sm.add_constant(data["educ"])
model_partial = sm.OLS(data["age"], X_partial).fit()
residuals_age = model_partial.resid

#3
model_residual = sm.OLS(y, residuals_age).fit()
beta1_residual = model_residual.params[0]

print(f"β1 from full model: {beta1_full}")
print(f"β1 from residual model: {beta1_residual}")

β1 from full model: 4.623447655158069
β1 from residual model: 4.623447655160652


<ipython-input-10-bd6d84255388>:14: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  beta1_residual = model_residual.params[0]


## (d)

In [11]:
X = data[["age", "smoke", "male"]]
y = data["birthwght"]

X_const = sm.add_constant(X)
model = sm.OLS(y, X_const)
results = model.fit(cov_type="HC0")
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:              birthwght   R-squared:                       0.040
Model:                            OLS   Adj. R-squared:                  0.037
Method:                 Least Squares   F-statistic:                     14.79
Date:                Sat, 07 Dec 2024   Prob (F-statistic):           1.98e-09
Time:                        10:20:42   Log-Likelihood:                -7696.2
No. Observations:                1000   AIC:                         1.540e+04
Df Residuals:                     996   BIC:                         1.542e+04
Df Model:                           3                                         
Covariance Type:                  HC0                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const       3299.5333     97.207     33.943      0.0

> 結果の解釈；smokeの回帰係数-253.4347は、喫煙する母親と喫煙しない母親とでは出生体重が約253g異なることを示している。この結果は,P値が0.000であることから強く支持することができる。また、maleの結果から男児の出生体重が女児より有意に高いことがわかる。しかし、ageに関してはP値から統計的に有意であるとは言えない。決定係数を見てみると、R-squared:0.040 であり低いため、より多くの説明変数を追加する必要がある。